feat(storage): support bucket restore in soft delete phase 2

mahendra-google opened the pull request:
Feature request includes the following:-

1. Get Bucket Generation
2. Get a Soft Deleted Bucket (Also soft-delete time and hard-delete time)
3. List Soft Deleted Buckets
4. Restore a Soft Deleted Bucket

jskeet commented on Directory.Packages.props:
```diff
@@ -1,5 +1,6 @@
 <Project>
   <ItemGroup>
+    <PackageVersion Include="Google.Apis.Storage.v1" Version="1.68.0.3604" />
```
Please revert this change, edit apis.json to change the dependency version (in the block for Google.Cloud.Storage.V1) and run generateprojects.sh to regenerate the project file.

mahendra-google replied:
Ok. I will revert this change and will edit apis.json with the same dependency version for (Google.Cloud.Storage.V1).

mahendra-google replied:
@jskeet  I have reverted my changes from Directory.Packages.props and updated Google.Apis.Storage.v1 version (1.68.0.3604) in apis.json file and after running generateprojects.sh , Google.Apis.Storage.v1 project file got automatically updated.

jskeet reviewed (commented):
This is not a full code review by any means, but I'm hoping that this little bit will reduce the number of iterations required after Amanda is back from Thanksgiving.

mahendra-google commented:
> This is not a full code review by any means, but I'm hoping that this little bit will reduce the number of iterations required after Amanda is back from Thanksgiving.

Ok

jskeet commented:
Unassigning myself from review - I'm on vacation until Tuesday. (I expect Amanda will review before then.)

amanda-tarafa commented:
I'll review today. @mahendra-google if you can rebase on main to remove conflicts, that'd be great, thanks.

mahendra-google commented:
> I'll review today. @mahendra-google if you can rebase on main to remove conflicts, that'd be great, thanks.

>@amanda-tarafa I have removed conflict for Directory.Packages.props file.

mahendra-google commented:
> > I'll review today. @mahendra-google if you can rebase on main to remove conflicts, that'd be great, thanks.
> 
> > @amanda-tarafa I have removed conflict for Directory.Packages.props file.

>> @amanda-tarafa I have removed project file conflict.

mahendra-google commented:
Storage Samples requires latest package of Google.Cloud.Storage.V1 having properties of  GetBucketOptions , ListBucketOptions, RestoreBucketOptions to work sample as expected. So for samples can I proceed with lastest dlls?

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/StorageFixture.cs:
```diff
@@ -60,6 +60,21 @@ public sealed class StorageFixture : CloudProjectFixtureBase, ICollectionFixture
         /// for 24 hours.
         /// </summary>
         public string SoftDeleteBucket => BucketPrefix + "-soft-delete";
+        /// <summary>
```
In the fixture, we only add resources that are going to be used in multiple tests. That's not the case for any of these. You should create, and clean up, these resources on the tests that need them. You should use there the fixture's CreateBucket method, etc.

mahendra-google replied:
Fixture's GenerateBucketName, CreateBucket methods are addded on the tests as recommended.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,44 @@
+            // We get bucket object using get bucket async method for soft delete bucket before deleting it. 
+            var bucket = await _fixture.Client.GetBucketAsync(_fixture.SoftDeleteBucketThree, new GetBucketOptions { SoftDeleted = false });
+            // We delete soft delete bucket.
+            try
```
Once you create the bucket on the test, you will know for a fact that it will be empty.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,44 @@
+                await _fixture.Client.DeleteBucketAsync(_fixture.SoftDeleteBucketThree);
+            }
+
+            // And now we can restore bucket using generation id.
```
In general avoid living comments that simply describe the code that follows, unless it's very complicated or particular code.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,26 @@ public void PartialResponses()
+            
+            foreach (var bucket in actualBuckets)
+            {
+                // Check if list contains only soft deleted buckets created in storage fixture
```
This is not what you are checking here. You are only saying to tests assertions on buckets whose names en with soft-delete.

Ain't there a metadata field indicating that the bucket has been softdeleted?

mahendra-google replied:
SoftDeleted option is there for ListBucketsOptionsto check if bucket has been softdeleted or not.

amanda-tarafa replied:
There are several things wrong with this test
- When this test executes there's absolutely no guaratee that you have run a test that has created a soft delete bucket. 
- You want to test that setting the SoftDeleteOnly option works, so you cannot rely on it working to test it works (as per your comment: "I have added SoftDeletedOnly metatdata for ListBucketsOptions to check if bucket has been softdeleted or not.")
- You are not testing that only soft-deleted buckets were returned, you are testing that if a bucket whose name ends in "soft-delete" is returned, then: its name is not null, its generation is not null, it has a soft delete time, and a hard delete time. If this test is passing anywhere it's because there's no "soft-deleted" bucket that has been created.

mahendra-google replied:
Changes are made in ListBucketTest.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+        catch (Exception)
+        {
+            // If bucket is not empty, we delete on a best effort basis.
```
Same comment as in the other test. You won't need this one you create the bucket in the test.

(And also there's an option that can be passed to indicate "delete the bucket and all its contents", so you really wouldn't even need this).

mahendra-google replied:
Option to delete  all its contents while deleting the bucket is passed  here and everywhere

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,84 @@
+            }
+            await _fixture.Client.DeleteBucketAsync(_fixture.SoftDeleteBucketOne);
+        }
+        // We get softDeleted object using get bucket async method for soft deleted bucket. 
```
You are not getting an object.

mahendra-google replied:
Comment removed

jskeet commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketOptions.cs:
```diff
@@ -52,6 +52,17 @@ public sealed class GetBucketOptions
+        public long? Generation { get; set; }
+
+        /// <summary>
+        /// (Optional) Set to true to retrieve a soft-deleted bucket .It will return the bucket metadata only if the bucket exists and is in a soft-deleted state.
```
Nit while I spot it: the period after the first sentence is in the wrong place.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+        // We get softDeleted object using get bucket async method for soft deleted bucket. 
+        var softDeleted = await _fixture.Client.GetBucketAsync(_fixture.SoftDeleteBucketOne, new GetBucketOptions { SoftDeleted = true, Generation = bucket.Generation });
+        Assert.NotNull(bucket.Generation);
+        Assert.NotNull(softDeleted.Generation);
```
I'm not sure what these are meant to be testing, shouldn't you test they are the same value?

mahendra-google replied:
Yes also bucket generations before and after delete are not null.

amanda-tarafa replied:
You can remove the null check then.

mahendra-google replied:
Null check for generation is removed

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    [Fact]
+    public async Task GetSoftDeletedBucket()
```
What's the difference between this test and the one above? What are the different cases they are testing?

mahendra-google replied:
The above test is for checking generation explicitly and this test for checking bucket name along with soft delete and hard delete time.

amanda-tarafa replied:
We don't need two tests for that.

mahendra-google replied:
I will share design doc let me know so I will add everything in single test

amanda-tarafa replied:
Yes, everything in a single test please.

mahendra-google replied:
Ok I will add in single test

mahendra-google replied:
Single test is kept and removed other one

amanda-tarafa reviewed (changes requested):
You need to add unit tests for all the option classes that you modified. See #12060 for examples.

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,44 @@
+using System;
```
Added license

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/ListBucketsOptions.cs:
```diff
@@ -61,7 +61,10 @@ public sealed class ListBucketsOptions
+        /// <summary>
+        /// (Optional) Set to true to list a soft-deleted buckets. It will return the bucket's metadata only if buckets exist and are in a soft-deleted state.
+        /// </summary>
+        public bool? SoftDeleted { get; set; }
```
SoftDeleted property changed to SoftDeletedOnly along with documentation.

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.RestoreBucket.cs:
```diff
@@ -0,0 +1,40 @@
+    private BucketsResource.RestoreRequest CreateRestoreBucketRequest(string bucket, long generation, RestoreBucketOptions options)
+    {
+        ValidateBucketName(bucket);
+        var request = Service.Buckets.Restore(bucket, generation);
```
I have added options to modify the request

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.RestoreBucket.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        ValidateBucketName(bucket);
+        var request = Service.Buckets.Restore(bucket, generation);
+        return request;
```
I have code for retry strategy

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketOptions.cs:
```diff
@@ -52,6 +52,17 @@ public sealed class GetBucketOptions
         public RetryOptions RetryOptions { get; set; }
 
+        /// <summary>
+        /// The bucket generation must be set (to retrieve a soft-deleted bucket) if soft_deleted is set to true.
```
```suggestion
        /// The bucket generation to be retrieved. It must be set if <see ref="SoftDelete" /> is true.
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketOptions.cs:
```diff
@@ -52,6 +52,17 @@ public sealed class GetBucketOptions
+        public long? Generation { get; set; }
+
+        /// <summary>
+        /// Set to true to retrieve a soft-deleted bucket. It will return the bucket metadata only if the bucket exists and is in a soft-deleted state.
+        /// The bucket generation must be set if soft_deleted is set to true.
```
Let's use something closer to the API documentaiton:

```suggestion
        /// If true, the soft-deleted version of the bucket will be retrieved.
        /// If true, <see ref="Generation"  /> must be set.
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,43 @@
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Storage.V1.IntegrationTests
```
For new files, file-scoped namespace please.

mahendra-google replied:
File-scoped namespace added

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,43 @@
+        public async Task RestoreSoftDeletedBucket()
+        {
+            var softDeleteBucket = _fixture.CreateBucket(Guid.NewGuid().ToString() + "-soft-delete", false, true);
+            var bucket = await _fixture.Client.GetBucketAsync(softDeleteBucket.Name, new GetBucketOptions { SoftDeleted = false });
```
You don't need this, this should return the same metadata that you already got when you created the bucket.

mahendra-google replied:
Removed line number 35 and used the same metadata while restoring bucket that I got when created bucket

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -90,5 +109,7 @@ private void AssertBucketNames(IEnumerable<Bucket> actualBuckets, string[] expec
             Assert.Equal(expectedNames.OrderBy(x => x), actualNames);
         }
+
+
```
Undo these two lines

mahendra-google replied:
2 blank lines removed

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,25 @@ public void PartialResponses()
         }
 
+        [Fact]
+        public async Task ListSoftDeletedBuckets()
```
```suggestion
        public async Task SoftDeletedOnly()
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,58 @@
+    public async Task GetBucketGeneration()
+    {
+        var softDeleteBucket  = _fixture.CreateBucket(Guid.NewGuid().ToString() + "-soft-delete", false, true);
+        var bucket = await _fixture.Client.GetBucketAsync(softDeleteBucket.Name, new GetBucketOptions { SoftDeleted = false });
```
You have the metadata already from having created the bucket. You don't need this.

mahendra-google replied:
I was using this to get and pass generation  in line 37 for GetBucketAsync menthod

amanda-tarafa replied:
I'm pretty sure you have the generation already in `softDeleteBucket`. The metadata returned on bucket creation should be the same as the one you are obtaining here, as the bucket shouldn't have changed because it's owned by this test.

mahendra-google replied:
Yes true  , generation from **softDeleteBucket** bucket is now passed while getting soft deleted bucket

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,58 @@
+    [Fact]
+    public async Task GetBucketGeneration()
+    {
+        var softDeleteBucket  = _fixture.CreateBucket(Guid.NewGuid().ToString() + "-soft-delete", false, true);
```
Please use the _fixture.GenerateBucketName method to generate the bucket names here and elsewhere.

mahendra-google replied:
_fixture.GenerateBucketName method for generating bucket name only? _fixture.CreateBucket is required for creating soft deleted bucket. Is that correct?

amanda-tarafa replied:
Yes, you still need to create the bucket using CreateBucket, but you should generate de bucket name using GenerateBucketName. (here and everywhere)

mahendra-google replied:
Bucket name is generated using **GenerateBucketName** and Bucket is created using **CreateBucket** (here and everywhere)

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,58 @@
+
+        await _fixture.Client.DeleteBucketAsync(softDeleteBucket.Name, new DeleteBucketOptions { DeleteObjects = true });
+        var softDeleted = await _fixture.Client.GetBucketAsync(softDeleteBucket.Name, new GetBucketOptions { SoftDeleted = true, Generation = bucket.Generation });
+        Assert.NotNull(bucket.Generation);
```
You don't need this assert, we are no testing GetObject.

mahendra-google replied:
Assert is removed to test GetObject

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,58 @@
+        var softDeleted = await _fixture.Client.GetBucketAsync(softDeleteBucket.Name, new GetBucketOptions { SoftDeleted = true, Generation = bucket.Generation });
+        Assert.NotNull(bucket.Generation);
+        Assert.NotNull(softDeleted.Generation);
+        Assert.Equal(bucket.Generation, softDeleted.Generation);
```
Can you also check that the soft delete timespan is set, that is the real confirmation that this is a soft deleted bucket.

mahendra-google replied:
As per design doc  I have added soft delete timespan in GetSoftDeletedBucket test. I will share design doc

mahendra-google replied:
Check for Soft delete timespan is set or not is added in single test along with other checks

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    [Fact]
+    public async Task GetBucketGeneration()
```
```suggestion
    public async Task SoftDeleted()
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/RestoreBucketOptionsTest.cs:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    [Fact]
+    public void ModifyRequest_MatchOptions()
```
```suggestion
    public void ModifyRequest_AllOptions()
```

JesseLovelace commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.RestoreBucket.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        ValidateBucketName(bucket);
+        var request = Service.Buckets.Restore(bucket, generation);
+        ApplyEncryptionKey(options?.EncryptionKey, kmsNameFromOptions: null, request);
```
Buckets can have a default encryption key, but that wouldn't be relevant to the restore bucket request. No need to support it as a RestoreBucketOption

mahendra-google replied:
Ok then I will remove code related to encryption support in restore bucket request.

mahendra-google replied:
Code removed related to encryption support in restore bucket request

JesseLovelace commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.RestoreBucket.cs:
```diff
@@ -0,0 +1,45 @@
+        var request = Service.Buckets.Restore(bucket, generation);
+        ApplyEncryptionKey(options?.EncryptionKey, kmsNameFromOptions: null, request);
+        options?.ModifyRequest(request);
+        RetryOptions retryOptions = options?.RetryOptions ?? RetryOptions.IdempotentRetryOptions;
```
Restore Bucket doesn't support the common idempotency params (ifGenerationMatch etc) but it is safely retryable, so it can probably use the same retry handler

mahendra-google replied:
Ok . I am keeping the same retry handler.

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RestoreBucketOptions.cs:
```diff
@@ -0,0 +1,58 @@
+    /// except for the bucket's access controls. This only affects
+    /// what information is returned when restoration is successful.
+    /// </summary>
+    public Projection? Projection { get; set; }
```
Blank line added between one property and the next

mahendra-google commented on apis/apis.json:
```diff
@@ -4967,7 +4967,7 @@
       "dependencies": {
         "Google.Api.Gax.Rest": "default",
-        "Google.Apis.Storage.v1": "1.68.0.3431"
+        "Google.Apis.Storage.v1": "1.68.0.3604"
```
@amanda-tarafa I have question. After rebasing to fork main merging with squash. Do I need to raise PR from main?

amanda-tarafa replied:
> After rebasing to fork main merging with squash

I don't understand what this means.

> Do I need to raise PR from fork main

You shouldn't need to create a new PR, you simply reabse this one on this repo (not your fork) main. So locally, and assuming this repo is configured as "upstream" (wheras your fork is configured as "origin"):

- fetch upstream: `git fetch upstream`
- interactively rebase this PR on upstream/main: `git rebase -i upstream/main`. Take the opportunity to squash the reviewed commits together, we don't care to include the review history as part of the repo history, once this PR is ready, it should be squashed into a single commit before merging.
- resolve the conflicts and complete the rebase operation
- force push the PR to your fork

mahendra-google replied:
Rebasing PR on upstream/main with squashing commits for already reviewed commits is completed also api version for Googl.Cloud.Storage.V1 is  changed in api.json file in generator-input folder

mahendra-google commented:
> You need to add unit tests for all the option classes that you modified. See #12060 for examples.

> Unit tests for all the option classes are added that I have modified

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.csproj:
```diff
@@ -9,8 +9,8 @@
     <PackageReference Include="ConfigureAwaitChecker.Analyzer" PrivateAssets="All" />
-    <PackageReference Include="Google.Api.Gax.Rest" />
-    <PackageReference Include="Google.Apis.Storage.v1" VersionOverride="[1.68.0.3431, 2.0.0.0)" />
+    <PackageReference Include="Google.Api.Gax.Rest" VersionOverride="[4.9.0, 5.0.0)" />
```
Changes are made

amanda-tarafa commented:
Thanks, I'll do a pass tomorrow.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/RestoreBucketTest.cs:
```diff
@@ -0,0 +1,42 @@
+    public async Task RestoreSoftDeletedBucket()
+    {
+        var bucketName = _fixture.GenerateBucketName();
+        var softDeleteBucket = _fixture.CreateBucket(bucketName, false, true);
```
```suggestion
        var softDeleteBucket = _fixture.CreateBucket(bucketName, multiVersion: false, softDelete: true);
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/ListBucketsOptionsTest.cs:
```diff
@@ -42,14 +43,16 @@ public void ModifyRequest_AllOptions()
             Assert.Equal(ProjectionEnum.Full, request.Projection);
             Assert.Equal("nextpage", request.PageToken);
             Assert.Equal("items(name),nextPageToken", request.Fields);
+            Assert.Equal(true, request.SoftDeleted);
```
```suggestion
            Assert.True(request.SoftDeleted);
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,27 @@ public void PartialResponses()
+        public async Task SoftDeletedOnly()
+        {
+            var bucketName = _fixture.GenerateBucketName();
+            var softDeleteBucket = _fixture.CreateBucket(bucketName, false, true);
```
Also here add the parameter names so the test is clearer.

mahendra-google replied:
parameter names so the test will be clearer

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,27 @@ public void PartialResponses()
+            await _fixture.Client.DeleteBucketAsync(softDeleteBucket.Name, new DeleteBucketOptions { DeleteObjects = true });
+            var actualBuckets = await _fixture.Client.ListBucketsAsync(_fixture.ProjectId, new ListBucketsOptions { SoftDeletedOnly = true }).ToListAsync();
+
+            foreach (var bucket in actualBuckets)
+            {   // Verify if the bucket is soft-deleted only 
+                Assert.NotNull(bucket.Generation);
+                Assert.NotNull(bucket.SoftDeleteTimeDateTimeOffset);
+                Assert.NotNull(bucket.HardDeleteTimeDateTimeOffset);
+
+                if (bucket.Name == softDeleteBucket.Name)
+                {   // Compare the generation number
+                    Assert.Equal(bucket.Generation, softDeleteBucket.Generation);
+                }
+            }
```
Please use Assert.All and Assert.Contains instead of a loop here.

mahendra-google replied:
Is it not working as expected ? because these changes are approved by My team's silver language leads using the Internal PR approval process.

amanda-tarafa replied:
There are different ways to achieve the same result, some are clearer than others. Please use Assert.All and Asssert.Contains.

amanda-tarafa replied:
Actually, is not only about clarity, this test is not sufficient as it is. It will pass even if `actualBuckets` is empty or if it does not contains `softDeleteBucket`. Using Assert.All and Assert.Contains will make the test sufficient.

mahendra-google replied:
If ```actualBuckets``` is made empty (``` actualBuckets = new List<Bucket>(); ```), count will be zero , it will not go inside **foreach** loop , So it will not pass , I would rather say it will skip assert.

mahendra-google replied:
@amanda-tarafa On the same lines , if I don't create a soft delete bucket in test itself (which I already created if I remove that piece of code of creation of soft delete bucket ) and hypothetically assuming storage fixture has also not created any soft delete bucket in any other tests , According to me in that case also `` actualBuckets `` count will be zero and it will not go inside loop. So it will skip asserts. Does that mean test should fail rather skip ?

amanda-tarafa replied:
A test that executes no asserts, assuming there are no exceptions (none in this case), is a passing test, it's not a skipped test. A passing test indicates that the feature has been tested and everything is working as expected, whereas a skipped test indicates that nothing was tested. This test is wrong because it may pass even when nothing is tested.

Explicitly, if there is a bug in production code that has the effect of the list returning empty (even when you have just created a soft deleted bucket), this test will pass, when it shouldn't have. That's because this test is saying:

- Create a soft deleted bucket (correct, we guarantee there's at least one soft deleted bucket so we can test).
- Fetch soft deleted buckets (the feature we want to test).
- If there happens to be elements in the list, make sure they are all soft deleted. (hmm, if the feature we want to test is broken so that the list is empty, then we really don't test anything and we don't find out the feature is broken)
- If the recently created bucket happens to be on the list, make sure all its fields have the expected values and that it is indeed soft deleted. (hmm, if the feature we want to test is broken so that the list is empty, then we really don't test anything and we don't find out the feature is broken)

The key being the "if it happens" where a test needs to make certain "it happened", no ifs and no buts.

What this test needs to do is:

- Create a soft deleted bucket
- Fetch soft deleted buckets
- Make sure the recently created bucket is on the list, and all its fields have the expected values. 
- Make sure all the elements on the list are indeed soft deleted.  Apart from the test own bucket, these may be buckets created by the fixture or by other tests, we don't care if there are none of these, or many, because we don't care about the order in which tests execute. The only thing that we care is that all tests in the list are soft deleted because that's the feature we are testing.

See those two last points describing the correct test, how they very naturally transalate to:

- Assert.Contains recently created bucket with the right fields.
- Assert.All are soft deletede buckets

You can certainly express the same semantic with a loop and some checks, but, it's way less naturally readable plus, errors like the one currently in this test are harder to spot and harder to explain, as this very thread demonstrates. Using the right assert statements makes the tests robust and correct more easily, plus way clearer.

Please fix.

mahendra-google replied:
Thanks for the elaborated reply it clears some of my doubts about the test as well  especially the production bug scenario.

mahendra-google replied:
Assert.All for soft deleted buckets and Asser.Contains for recently created bucket with metatdata are added in listbucket test

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketOptions.cs:
```diff
@@ -52,6 +52,17 @@ public sealed class GetBucketOptions
+
+        /// <summary>
+        /// If true, the soft-deleted version of the bucket will be retrieved.
+        /// If true, <see ref="Generation"  /> must be set.
```
```suggestion
        /// If true, <see ref="Generation" /> must be set.
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/GetBucketOptionsTest.cs:
```diff
@@ -41,13 +43,17 @@ public void ModifyRequest_PositiveMatchOptions()
             Assert.Null(request.IfMetagenerationNotMatch);
             Assert.Equal(ProjectionEnum.Full, request.Projection);
             Assert.Equal("proj", request.UserProject);
+            Assert.Equal(true, request.SoftDeleted);
```
```suggestion
            Assert.True(request.SoftDeleted);
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/GetBucketOptionsTest.cs:
```diff
@@ -61,6 +67,8 @@ public void ModifyRequest_NegativeMatchOptions()
             };
             options.ModifyRequest(request);
             Assert.Null(request.IfMetagenerationMatch);
+            Assert.Null(request.SoftDeleted);
+            Assert.Null(request.Generation);
```
We don't need to add these here, this test is very specifically testing for metageneration match and no match being alternates.

mahendra-google replied:
Null checks removed

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/GetBucketTest.cs:
```diff
@@ -0,0 +1,43 @@
+    public async Task SoftDeleted()
+    {
+        var bucketName = _fixture.GenerateBucketName();
+        var softDeleteBucket = _fixture.CreateBucket(bucketName, false, true);
```
Add the name of the parameters here.

mahendra-google replied:
name of parameters are added

amanda-tarafa reviewed (changes requested):
Almost there, thanks!

mahendra-google commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RestoreBucketOptions.cs:
```diff
@@ -0,0 +1,61 @@
+    /// The encryption key to use for this operation. If this property is null, the <see cref="StorageClient.EncryptionKey"/>
+    /// will be used instead. Use <see cref="EncryptionKey.None"/> to remove encryption headers from this request.
+    /// </summary>
+    public EncryptionKey EncryptionKey { get; set; }
```
Encryption key property is removed as per reference from @JesseLovelace [comment](https://github.com/googleapis/google-cloud-dotnet/pull/13894#discussion_r1919146580)

amanda-tarafa commented:
@mahendra-google Do let me know here when this is ready for review again. No rush on my part, but just know I won't review until you have said so here. Thanks!

mahendra-google commented:
> @mahendra-google Do let me know here when this is ready for review again. No rush on my part, but just know I won't review until you have said so here. Thanks!

>@amanda-tarafa Yes right , I will leave a comment for re-review in PR itself like @danielduhh left , apart from clicking on re-review icon post completion of internal PR review from my team as a part of process. Thanks!

amanda-tarafa commented:
> > @amanda-tarafa Yes right , I will leave a comment for re-review in PR itself like @danielduhh left , apart from clicking on re-review icon post completion of internal PR review from my team as a part of process. Thanks!

That's perfect, thanks!

amanda-tarafa commented:
@mahendra-google let me know if I can help with this one. Thanks.

mahendra-google commented:
> @mahendra-google let me know if I can help with this one. Thanks.

> @amanda-tarafa Yes.Thanks

mahendra-google commented:
This PR is ready for a re-review @amanda-tarafa

amanda-tarafa commented:
Thanks, I'll take a look tomorrow.

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,58 @@ public void PartialResponses()
+                )
+                .ToListAsync();
+
+            var recentlyCreatedSoftDeleteBucket = actualBuckets
+                 .Where(b => b.Name == softDeleteBucket.Name)
+                 .ToList();
+            Assert.Contains(
+                recentlyCreatedSoftDeleteBucket.First().Name.ToString(),
+                softDeleteBucket.Name.ToString()
+            );
+            Assert.Contains(
+                recentlyCreatedSoftDeleteBucket.First().Generation.ToString(),
+                softDeleteBucket.Generation.ToString()
+            );
+            Assert.NotNull(recentlyCreatedSoftDeleteBucket.First().SoftDeleteTimeDateTimeOffset);
+            Assert.NotNull(recentlyCreatedSoftDeleteBucket.First().HardDeleteTimeDateTimeOffset);
+
+            if (actualBuckets.Count == 0)
+            {
+                Assert.Fail("Soft deleted bucket list is empty");
+            }
+            else
+            {
+                Assert.All(actualBuckets, AssertSoftDeletedBucket);
+            }
```
This is how Assert.Contains and Assert.All are used in this test. I'll self push this suggestion.

```suggestion
            // Check the list cotains the bucket we just soft-deleted.
            Assert.Contains(actualBuckets, bucket => bucket.Name == softDeleteBucket.Name && bucket.Generation == softDeleteBucket.Generation);
            // Check all the buckets in the list are soft-deleted buckets.
            Assert.All(actualBuckets, AssertSoftDeletedBucket);
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,58 @@ public void PartialResponses()
+            }
+        }
+
+        // Validating whether all buckets are soft-deleted only.
```
```suggestion
        // Validates that the given bucket is soft-deleted.
```

amanda-tarafa commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/ListBucketsTest.cs:
```diff
@@ -75,6 +75,58 @@ public void PartialResponses()
+        public async Task SoftDeletedOnly()
+        {
+            var bucketName = _fixture.GenerateBucketName();
+            var softDeleteBucket = _fixture.CreateBucket(
+                bucketName,
+                multiVersion: false,
+                softDelete: true
+            );
+            await _fixture.Client.DeleteBucketAsync(
+                softDeleteBucket.Name,
+                new DeleteBucketOptions { DeleteObjects = true }
+            );
+            var actualBuckets = await _fixture
+                .Client.ListBucketsAsync(
+                    _fixture.ProjectId,
+                    new ListBucketsOptions { SoftDeletedOnly = true }
+                )
+                .ToListAsync();
```
Lots of unnecessary wraping here:

```suggestion
            var softDeleteBucket = _fixture.CreateBucket(bucketName, multiVersion: false, softDelete: true);
            
            await _fixture.Client.DeleteBucketAsync(softDeleteBucket.Name, new DeleteBucketOptions { DeleteObjects = true });
            
            var actualBuckets = await _fixture.Client
                .ListBucketsAsync(_fixture.ProjectId, new ListBucketsOptions { SoftDeletedOnly = true })
                .ToListAsync();
```

amanda-tarafa reviewed (commented):
I've made suggestions on how Assert.Contains is meant to be used in this case, and remove excesive line wrapping. I'll push these changes myself and approve the PR after the tests are green.

amanda-tarafa reviewed (approved):
Tests are green after my changes. Approving and merging.
